userspace/chisel/lua_parser.cpp

```diff
@@ -39,50 +38,29 @@ const static struct luaL_reg ll_filter [] =
 	{NULL,NULL}
 };
 
-lua_parser::lua_parser(gen_event_filter_factory &factory, lua_State *ls, const char *lua_library_name)
-	: m_factory(factory)
+lua_parser::lua_parser(std::shared_ptr<gen_event_filter_factory> factory)
+	: m_factory(factory), m_last_boolop(BO_NONE),
+	  m_have_rel_expr(false), m_nest_level(0)
 {
-	m_filter = NULL;
-
-	m_ls = ls;
-	reset();
-
-	// Register our c++ defined functions
-	luaL_openlib(m_ls, lua_library_name, ll_filter, 0);
+	m_filter.reset(m_factory->new_filter());
 }
 
-void lua_parser::reset()
+lua_parser::~lua_parser()
 {
-	m_have_rel_expr = false;
-	m_last_boolop = BO_NONE;
-	m_nest_level = 0;
-
-	m_filter = m_factory.new_filter();
 }
 
-gen_event_filter* lua_parser::get_filter(bool reset_filter)
+void lua_parser::register_callbacks(lua_State *ls, const char *lua_library_name)
 {
-	if (m_nest_level != 0)
-	{
-		throw sinsp_exception("Error in configured filter: unbalanced nesting");
-	}
-
-	gen_event_filter *ret = m_filter;
-
-	if (reset_filter)
-	{
-		reset();
-	}
-
-	return ret;
+	// Register our c++ defined functions
+	luaL_openlib(ls, lua_library_name, ll_filter, 0);
 }
-lua_parser::~lua_parser()
-{
-	// The lua state is not considered owned by this object, so
-	// not freeing it.
 
-	delete m_filter;
-	m_filter = NULL;
+std::shared_ptr<gen_event_filter> lua_parser::filter()
+{
+	return m_filter;
 }
 
-
+std::shared_ptr<gen_event_filter_factory> lua_parser::factory()
+{
+	return m_factory;
+}
```

Comment:
Can we return `std::shared_ptr<T>&` and maybe save a refcount inc/dec? No idea whether it's worth it so your call

Reply:
I prefer to not play games like that for shared_ptrs and just let the refcounting do it's thing. This isn't performance-sensitive really, as it only happens when compiling filters, which is relatively uncommon.